ASC-393 Refactoring of proxy_test.go

jkopriva opened the pull request:
This PR about refactoring proxy_test.go file and splitting it in more files to make tests readable.

Parallelization of proxy tests will be part of next PRs.

xcoulon commented on test/e2e/proxy/applications_test.go:
```diff
@@ -0,0 +1,352 @@
+					patchPayloadBytes, err := json.Marshal(patchPayload)
+					require.NoError(t, err)
+
+					// Appply Patch
```
```suggestion
					// Apply Patch
```

jkopriva replied:
Thank you! Fixed. Resolving.

xcoulon commented on test/e2e/proxy/applications_test.go:
```diff
@@ -0,0 +1,352 @@
+				resp, err = client.Do(request)
+				require.NoError(t, err)
+				defer resp.Body.Close()
+				var body []byte
+				body, err = io.ReadAll(resp.Body)
```
```suggestion
				body, err := io.ReadAll(resp.Body)
```

jkopriva replied:
Thank you! Fixed. Resolving.

xcoulon reviewed (commented):
contents looks good overall! I just left a few comments about the form

jkopriva commented on test/e2e/proxy/applications_test.go:
```diff
@@ -0,0 +1,352 @@
+// To work around this limitation, we created a dummy HAS API type that has the same name
+// and the same group as the actual one. The CRD is created as part of the test setup
+// and since the CRD name & group name matches, then RBAC allow us to execute create/read
+// operations on that resource using the user permissions.
```
Thank you! Resolving.

jkopriva commented on test/e2e/proxy/applications_test.go:
```diff
@@ -0,0 +1,352 @@
+
+				request.Header.Set("Authorization", fmt.Sprintf("Bearer %s", user.Token))
+				var resp *http.Response
+				resp, err = client.Do(request)
```
Thank you! Fixed. Resolving.

jkopriva commented on test/e2e/proxy/applications_test.go:
```diff
@@ -0,0 +1,352 @@
+				}
+
+				t.Run("use proxy to update a HAS Application CR in the user appstudio namespace via proxy API", func(t *testing.T) {
+					// Update application
```
Should be fixed now. Resolving.

rajivnathan commented on test/e2e/proxy/applications_test.go:
```diff
@@ -0,0 +1,361 @@
+// and the same group as the actual one. The CRD is created as part of the test setup
+// and since the CRD name & group name matches, then RBAC allow us to execute create/read
+// operations on that resource using the user permissions.
+func TestProxyApplicationsFlow(t *testing.T) {
```
```suggestion
func TestProxyFlow(t *testing.T) {
```
Even though the test is creating Application resources it's still mainly about the proxy flow itself. ie. Testing that the request to the proxy is transmitted to the member cluster's api server correctly, etc. So I think the `TestProxyFlow` name was more accurate.

jkopriva replied:
Fixed, thanks!

rajivnathan reviewed (commented):
Thanks, these are great changes! They look good overall. I've left a few comments about the code but I have some general questions/comment as well:

1. If I understand the changes correctly, it's mainly taking everything that was in proxy_test.go and splitting it into separate files and tests and making the tests run in parallel. So shouldn't we delete the proxy_test.go file as part of this PR?

2. I think we would need to move the test files from `test/e2e/proxy` to `test/e2e/parallel/proxy` in order for them to run in parallel since only tests under the `test/e2e/parallel` dir are run in parallel. See https://github.com/codeready-toolchain/toolchain-e2e/blob/2d619cfee45849678d9691a1ce95e6f3b8386ffb/make/test.mk#L119C186-L119C186

3. Would it be better to rename the `testsupport/appstudio` package to `testsupport/proxy`? I think those utilities have more to do with the proxy than appstudio.

MatousJobanek commented on testsupport/appstudio/watcher.go:
```diff
@@ -0,0 +1,251 @@
+	// this one is the expected timeout and should be bigger than the default one that was originally set
+	// for OpenShift Route and the RoundTripper inside proxy to make sure that the call is terminated
+	// via the context and not by the server.
+	contextTimeout := 40 * time.Second
```
let's keep using this specific timeout - the reasons for that:
1. this is not a timeout for which we would wait for some result. This is to verify that the watch call is not terminated in the middle by a wrong configuration either in a Route or in the Proxy
2. the test waits until the watch call is terminated after 40 seconds. We don't want to wait for 2 minutes and make the tests even longer
3. This specific context variable contains the description why and how is it being used

jkopriva replied:
Hi @MatousJobanek , thank you for context. I will keep it as it is. Resolving.

MatousJobanek commented:
Hey @jkopriva, thanks a lot for your PR.
I haven't reviewed the content of the PR yet, just two thoughts before doing that:

**Test execution**
You created a new package/directory - make sure that the tests in the directory are being executed. I don't see them in the build log. My guess is that you need to modify the makefile target so it runs tests also in sub-packages. Correct me if I'm wrong

**PR description**
Please add some description to the PR so we understand what you have done here, why, and what is the plan of using it in the future PRs, so we can review that the intended changes are done. For example, Rajiv mentioned:
>  If I understand the changes correctly, it's mainly taking everything that was in proxy_test.go and splitting it into separate files and tests and making the tests run in parallel. 
...
I think we would need to move the test files from test/e2e/proxy to test/e2e/parallel/proxy in order for them to run in parallel since only tests under the test/e2e/parallel dir are run in parallel. See https://github.com/codeready-toolchain/toolchain-e2e/blob/2d619cfee45849678d9691a1ce95e6f3b8386ffb/make/test.mk#L119C186-L119C186

But I don't see any parallelization there yet, or am I missing something?

jkopriva commented on testsupport/appstudio/proxy_user.go:
```diff
@@ -0,0 +1,164 @@
+	CompliantUsername     string
+}
+
+func CreateAppStudioUser(t *testing.T, awaitilities wait.Awaitilities, user *ProxyUser) {
```
Fixed, thanks!

jkopriva commented on test/e2e/proxy/space_listener_test.go:
```diff
@@ -0,0 +1,238 @@
+package proxy
```
Fixed, thanks!

jkopriva commented on testsupport/appstudio/utils.go:
```diff
@@ -0,0 +1,42 @@
+	}
+}
+
+// SetAppstudioConfig applies toolchain configuration for stone soup scenarios
```
Fixed, thanks!

jkopriva commented on testsupport/appstudio/utils.go:
```diff
@@ -0,0 +1,42 @@
+
+// SetAppstudioConfig applies toolchain configuration for stone soup scenarios
+func SetAppstudioConfig(t *testing.T, hostAwait *wait.HostAwaitility, memberAwait *wait.MemberAwaitility) {
+	// member cluster configured to skip user creation to mimic stonesoup configuration where user & identity resources are not created
```
Fixed, thanks!

jkopriva commented:
Hi @rajivnathan, Thank you for review!
ad 1. Its deleted now(it get back during rebase).
ad 2. I will add the parallelization in next PR(s). This was mostly for refactoring the proxy_test.go file to have it more readable.
ad 3. Moved/renamed.

jkopriva commented:
@MatousJobanek Hi, thank you for comments.
Ad Tests execution
You are correct with I have updated the make file to also include "new" proxy tests. 
(I was thinking maybe revisiting make files could be done also in future)
AD PR description
You are correct, I should have added better PR description and what is the plan, I will do it in next PRs.
(I have updated the description in this PR)

MatousJobanek commented on test/e2e/proxy/applications_test.go:
```diff
@@ -0,0 +1,361 @@
+// and the same group as the actual one. The CRD is created as part of the test setup
+// and since the CRD name & group name matches, then RBAC allow us to execute create/read
+// operations on that resource using the user permissions.
+func TestProxyFlow(t *testing.T) {
```
related to the other comment, I don't see how we could get closer to sending many requests to proxy in a parallel way. There are many sub-tests that will be executed in sequential order.
I hoped that we could have something closer to what we do in migration tests. see 
https://github.com/codeready-toolchain/toolchain-e2e/blob/713b9018786877b8782f3aecd0c3744616d50382/test/migration/setup_runner.go#L38-L59
https://github.com/codeready-toolchain/toolchain-e2e/blob/713b9018786877b8782f3aecd0c3744616d50382/test/migration/verify/verify_migration_test.go#L68-L106

A very very quick example of what it could look like:
```go
...
usersToBeTested := createUsersForParallelTests("user1"..."user10")
usersToBeTested["user1"].ShareSpaceWith(t, hostAwait, users["user2"])
usersToBeTested["user4"].ShareSpaceWith(t, hostAwait, users["user2"])
usersToBeTested["user10"].ShareSpaceWith(t, hostAwait, users["user5"])

...

for _, user := range usersToBeTested {
	// provision users
	user.provision()
	user.shareWithOtherUsers()
	user.verify().assessToWorkspaces() // space lister verification

	for _, action := actionsToBeTested { // CRUD, watch, websockets operations, etc...
		go func() {
			user.verify().action(action).inAllWorkspaces()
		}()
	}
}
...
```
the actions to be tested are the whole stories like
action 1:
1. create CM
2. list CMs
3. delete CM

action2
1. watch secrets
2. create secret
3. modify it
4. delete it

action3
1. create pod
5. see the logs
6. delete pod

etc...

All these things could be easily hidden behind a nice fluent API and set of functions. You can technically reuse most of the functions that are already there - eg the one for running websockets.

Is this still your plan and this PR is only a preparation for it?

jkopriva replied:
Thank you for this comment. I am working on running tests in parallel in similar way, as you pointed out and this PR is more about preparation for next PR(s) with parallelization.

MatousJobanek reviewed (commented):
overall looks good - If I understand the changes correctly, it's just about moving the tests and functions around and separating them. 
I'm slightly missing the point of introducing a new package and moving the tests there and I'm also missing your plans for introducing parallelization. Could you please add a topic to the next sync call so we can discuss it?

jkopriva commented on testsupport/proxy/utils.go:
```diff
@@ -0,0 +1,42 @@
+}
+
+// SetAppstudioConfig applies toolchain configuration for appstudio scenarios
+func SetAppstudioConfig(t *testing.T, hostAwait *wait.HostAwaitility, memberAwait *wait.MemberAwaitility) {
+	// member cluster configured to skip user creation to mimic appstudio configuration where user & identity resources are not created
+	memberConfigurationWithSkipUserCreation := testconfig.ModifyMemberOperatorConfigObj(memberAwait.GetMemberOperatorConfig(t), testconfig.SkipUserCreation(true))
+	// configure default space tier to appstudio
+	hostAwait.UpdateToolchainConfig(t, testconfig.Tiers().DefaultUserTier("deactivate30").DefaultSpaceTier("appstudio"), testconfig.Members().Default(memberConfigurationWithSkipUserCreation.Spec))
+}
```
This will be deleted in future - here it is just to avoid duplicated code.

MatousJobanek commented:
Let's close this PR for now.
The test case has been moved to the parallel test-suite in the meantime. Also, given the latest changes in the set of priorities and org structure, I don't think that you would have time to take a look at it soon. 
We can reopen it if the PR becomes valid again. 
But thanks a lot for your contribution and your effort @jkopriva
